### PR TITLE
D3D12: refactoring of kernel argument constant buffer packing

### PR DESCRIPTION
**Pull request by slomp**
The `halide_run()` interface with `arg_sizes[]` alone wasn't enough for safe argument packing; argument type information is needed.
NOTE: this work has been based on https://github.com/halide/Halide/pull/5544 : merge that before merging this.

### CODE REVIEW

File: src/runtime/d3d12compute.cpp
Comment (steven-johnson) [inline src/runtime/d3d12compute.cpp:3133]: Can a Bool (aka UInt(1)) ever be passed in here?

**Comment by steven-johnson** [comment]
LGTM pending green

**Comment by steven-johnson** [comment]
(but fix clang-format first)

**Comment by slomp** [comment]
> Can a Bool (aka UInt(1)) ever be passed in here?

Hmmm, that's a possibility... The previous code was sort of handling bool implicitly as uint8...
Let's see what the Win64 tests tell us.

> (but fix clang-format first)

Thanks, I rushed opening this PR so that Shoaib could have an early look, and I was gonna run-clang format later, but it looks like things are working well already.

**Comment by steven-johnson** [comment]
yeah, I ran clang-format myself so I could clear out semistale builds.

**Comment by slomp** [comment]
> Can a Bool (aka UInt(1)) ever be passed in here?

Ok, Win64 tests failed, I'm gonna handle `Bool` / `Uint(1)`. Just to be sure, it's never going to be signed `Int(1)`, right?

**Comment by steven-johnson** [comment]
LGTM

**Comment by slomp** [comment]
It looks like `lens_blur` failed with `DXGI_ERROR_DEVICE_REMOVED`.
Not sure if it was a fluke, or if it's a genuine crash (the reported GPU was the 750Ti, which has shown problems in the past).

**Comment by steven-johnson** [comment]
If we have a known-ancient-and/or-flaky GPU, we should replace it. (Hopefully the new Winbots will help with this.)

**Comment by slomp** [comment]
> If we have a known-ancient-and/or-flaky GPU, we should replace it. (Hopefully the new Winbots will help with this.)

For the record, the (potential) issue with the 750Ti is that it was released in early 2014, and D3D12 was released in mid-2015, so the support for it was back-ported. The card seems to have everything that the baseline D3D12 tier cards need, but I don't think enough attention has been directed towards its drivers.

**Comment by shoaibkamil** [comment]
I'll run the test in a loop overnight locally to see if it's reproducible here.

**Comment by slomp** [comment]
Looks like we're good, it was a red-herring from the 750Ti after all.
Just waiting for someone to approve to get it merged.

**Review by steven-johnson — APPROVED** [review]

**Pull request merged**
